### ISSUE TITLE
Allow blacklisting of domains via resolved

### ISSUE THREAD

**Issue by zabbal** [systemd/systemd#766]
That's the question which pop-up over and over again - people wonder why can't they use 
0.0.0.0 *.adexprt.com
instead of
0.0.0.0 cdn1.adexprt.com
0.0.0.0 cdn2.adexprt.com
0.0.0.0 cdn3.adexprt.com
0.0.0.0 cdn4.adexprt.com
...
in their /etc/hosts

Of course there's a workaround of running local caching dns server - but it seems kinda insane for such a trivial goal. Since you plan to get things fixed with resolved anyway I wonder if you could implement this as well?

For me it would be huge win which would help to reduce my /etc/hosts from 8Mb to some saner values :)

**Comment by zonque** [comment]
sd-resolved doesn't deal with `/etc/hosts`, that's up to the nss `files` module. So for now, there's not much we can do.

**Comment by zabbal** [comment]
I don't have particular emotional attachment to /etc/hosts :)
What I'm looking for is some place to configure wildcard dns for my local host (and maybe containers running on it) - it could be /etc/systemd/resolved.conf as well.

**Comment by zonque** [comment]
Let me rephrase then: we don't support any static host name databases in resolved at this time. Not sure whether that will ever happen. But I understand your motivation for the proposal, so we'll keep that in mind :)

**Comment by zabbal** [comment]
To add a bit more stress to it - in addition to ability to easily suppress junk I'm also concerned with cases like http://hackercodex.com/guide/how-to-stop-isp-dns-server-hijacking/ - having ability to lock down name resolution for couple of crucial services I use regardless of my greedy ISP actions won't solve the problem entirely but at least is a nicer workaround than nothing. So there's also security aspect to it.

**Comment by zonque** [comment]
Yes, that's what `/etc/hosts` is for, and as long as you order `files` before anything else in `/etc/nsswitch.conf`, that covers your case. It's just that we currently don't implement such functionality in sd-resolved.

**Comment by poettering** [comment]
Also, /etc/hosts is not a systemd configuration file, but defined by glibc. As such we should be very careful with extending its format, outside of what glibc defines.

If the functionality you ask for is something we want to support, then the right way would probably be via an explicit configuration option Blacklist= that takes a list of domain names and blocks them and everything beneath them, and returns a proper error code to clients, rather than a fake 0.0.0.0 reply.

**Comment by johannbg** [comment]
@poettering  Oh come on I was interested to see the outcome of what would happen if you would have accepted/merged/implemented something that collided with gethostbyaddr when it associates an IP address with a name ( which is supported which many are not aware of due to it not commonly being implemented and used ) as well as any other weird hard to trace dns issues and application bugs that would have been triggered when you would have implemented support for wildcards in host or resolved ;) 

I thought Daniel was just being polite not closing this WONTFIX immediately you know since it's quite obvious ( or that it is quite obvious that there was atleast some reason behind it ) why wildcard have never been "supported" in all those years ;)

**Comment by zabbal** [comment]
@johannbg - this is RFE not a bug hence it doesn't really matter that the feature did not exist before. Lots of things in systemd didn't. And I really doubt that having Blacklist=adexprt.com,doubleclick.net in /etc/systemd/resolved.conf would break anything for any of the existing gethostbyaddr() clients - it would be much easier to configure than stuffing all that data into /etc/hosts or configuring local dns server but it would be completely transparent to all the existing clients as long as they respect nss.

**Comment by johannbg** [comment]
I was perfectly aware that this was an RFE not a bug and what you need to do is to ask yourself why this has this not been implemented in either places ( hosts or resolved.conf ) in all those years ( I think verisign started hijacking dns responses in what 2003 ) which is what matters. . .

Blacklisting is best kept outside resolved and resolved.conf.  There exist other application you can use to be playing blacklisting ( which is either manually or automatically updated ) whack-a-mole game with ISP's. If the intent is to address this particular problem the solution that would need to be implemented in resolved would need to be smarter than that. . .

**Comment by johannbg** [comment]
@poettering @zonque when ye finally come to the conclusion that this problem can only be solved by implementing/support DNSSEC, the current status of that, was that it was in status qou for various upstreams both about what the "trusted nameserver" entry in resolv.conf was supposed to look like ( yes adding entries like wildcard or blacklist etc will make certain existing applications crap itself ) and whether DNSSEC belong or should be added to libc-API. 

You should be able to ping internally Petr Spacek ( since you all work for the same employee  ) whom was neck deep in code and discussions when they took place roughly year ago.

**Comment by teg** [comment]
I think we should not add blacklist support to resolved, and should close this RFE. If anything this is something to be done at a much higher level in the stack, depending on the usecase (adblocker in the browser for instance).

As to DNSSEC, yes, we need/want/are getting this, and that should solve things like hijacking.

**Comment by zabbal** [comment]
I think for dns-blacklisting the more logical place would be dns-related service like systemd-resolfved. The browser is for http-level blacklisting just like iptables is for ip-level filtering.

**Comment by poettering** [comment]
resolved in git master actually implements this, as it will read /etc/hosts now and synthesize RRs for it. It also recognizes 0.0.0.0 and ::0 specially, and make sure an NODATA reply is synthesized. This should basically get you what you want.

**Comment by zabbal** [comment]
Not really: the idea was to move from /etc/hosts which weights couple of megabytes due to constant repetition of entries for subdomains like: 
0.0.0.0 cdn1.adexprt.com
0.0.0.0 cdn2.adexprt.com
0.0.0.0 cdn3.adexprt.com
0.0.0.0 cdn4.adexprt.com
...
to a saner config file of systemd-resolved with single entry like:
0.0.0.0 *.adexprt.com
covering all possible subdomains via wildcard.

**Comment by poettering** [comment]
I am not convinced that such wildcarding is really something I'd like to see in resolved...
